signer/WebCryptoSigner.js

```diff
@@ -0,0 +1,111 @@
+const jsrsasign = require('jsrsasign');
+
+
+/**
+ * The signer based on Web Crypto API
+ */
+class WebCryptoSigner {
+  /**
+   * @param {String} pkcs1 PKCS#1
+   */
+  constructor(pkcs1) {
+    this.pkcs1 = pkcs1;
+  }
+
+  /**
+   * @param {Uint8Array} content
+   * @return {Uint8Array}
+   */
+  async sign(content) {
+    if (!this.pkcs8) {
+      this.pkcs8 = await this._PKCS1ToPKCS8(this.pkcs1);
+    }
+
+    const algorithm = { // EcdsaParams
+      name: 'ECDSA',
+      hash: 'SHA-256',
+    };
+    const data = content;
+    let key;
+
+    try {
+      key = await this._key(this.pkcs8);
+    } catch (_) {
+      throw new Error('Failed to sign the request');
+    }
+
+    try {
+      const signature = await window.crypto.subtle.sign(algorithm, key, data);
+      return this._P1363ToDer(new Uint8Array(signature));
+    } catch (_) {
+      throw new Error(`Failed to sign the request`);
+    }
+  }
+
+  /**
+   * @param {String} pkcs1
+   */
+  async _PKCS1ToPKCS8(pkcs1) {
+    pkcs1 = pkcs1.replace('-----BEGIN EC PRIVATE KEY-----', '')
+        .replace('-----END EC PRIVATE KEY-----', '')
+        .replace(/\n/g, '');
+    const k = jsrsasign.KEYUTIL.getKey(
+        jsrsasign.b64utohex(pkcs1), null, 'pkcs5prv'
+    );
+    const kk = new keyutil.Key('oct',
+        new Uint8Array(jsrsasign.hextoArrayBuffer(k.prvKeyHex)),
+        {namedCurve: 'P-256'}
+    );
+    return await kk.export('pem');
+  }
+
+  /**
+   * @param {Uint8Array} sig
+   * @return {Uint8Array}
+   */
+  _P1363ToDer(sig) {
```

Comment:
It looks pretty tricky. Is there any reference in this conversion algorithm ?

Reply:
@feeblefakie
It is from the sample code of [token.io](developer.token.io).
It follows the standard to do the conversion.

ECDSA generates two values, `r` (32 bytes) and `s` (26 bytes).

In ASN.1 DER, it will be encoded into HEX with the following sequence.
```
[
  0x30 (DER tag for sequence),
  length of r + s (32/33 + 26/27) + 2 (two tags) + 2 (length fields),
  0x02 (DER tag for integer),
  32 or 33,
  r,
  0x02 (DER tag for integer),
  26 or 27,
  s
]
```

Notice because DER only defines signed integer, so that if the first bit of r (or s) is not 0,
we need another padding byte.
That is the reason that the length of r might be 33 and s might be 27.

P1363 is very simple.
It is encoded in 64 bytes HEX with the following sequence.
```
[
  r (32 bytes)
  0x00,
  0x00,
  0x00,
  0x00,
  0x00,
  0x00,
  s (26 bytes)
]
```

Notice we need 6 extra padding bytes because s has only 26 bytes.

Reply:
Thanks !
It would be nice if you leave a comment on the code.

Reply:
Understood.